File: webhook-audit-blotter/node/src/signatureVerifier.ts
/**
 * HMAC-SHA256 signature check for an Evagene webhook delivery.
 *
 * The comparison uses `timingSafeEqual` to avoid leaking the secret via
 * wall-clock timing differences between a matching and a mismatching byte.
 * Never compare signatures with `===`.
 */

import { createHmac, timingSafeEqual } from 'node:crypto';

const HEX_PREFIX = 'sha256=';
const SHA256_HEX_LENGTH = 64;
const HEX_CHAR_RE = /^[0-9a-f]+$/i;

export function verifySignature(
  rawBody: Buffer,
  signatureHeader: string | undefined,
  secret: string,
): boolean {
  const presented = parseSignatureHeader(signatureHeader);
  if (presented === undefined) {
    return false;
  }
  const expected = createHmac('sha256', secret).update(rawBody).digest();
  return presented.length === expected.length && timingSafeEqual(presented, expected);
}

function parseSignatureHeader(header: string | undefined): Buffer | undefined {
  if (header === undefined) {
    return undefined;
  }
  const stripped = header.startsWith(HEX_PREFIX) ? header.slice(HEX_PREFIX.length) : header;
  if (stripped.length !== SHA256_HEX_LENGTH || !HEX_CHAR_RE.test(stripped)) {
    return undefined;
  }
  return Buffer.from(stripped, 'hex');
}

File: webhook-audit-blotter/python/src/webhook_audit_blotter/signature_verifier.py
"""HMAC-SHA256 signature check for an Evagene webhook delivery.

The comparison uses :func:`hmac.compare_digest` to avoid leaking the
secret via wall-clock timing differences between a matching and a
mismatching byte.  Never compare signatures with ``==``.
"""

from __future__ import annotations

import hmac
from hashlib import sha256

_HEX_PREFIX = "sha256="
_SHA256_HEX_LENGTH = 64


def verify_signature(raw_body: bytes, signature_header: str | None, secret: str) -> bool:
    presented = _parse_signature_header(signature_header)
    if presented is None:
        return False
    expected = hmac.new(secret.encode("utf-8"), raw_body, sha256).digest()
    return hmac.compare_digest(presented, expected)


def _parse_signature_header(header: str | None) -> bytes | None:
    if header is None:
        return None
    stripped = header[len(_HEX_PREFIX) :] if header.startswith(_HEX_PREFIX) else header
    if len(stripped) != _SHA256_HEX_LENGTH:
        return None
    try:
        return bytes.fromhex(stripped)
    except ValueError:
        return None
